ibtk/src/lagrangian/FEProjector.cpp

```diff
@@ -0,0 +1,426 @@
+                                M_vec->set(node_dof_index, 1.0);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // Assemble the vector.
+        M_vec->close();
+
+        // Store the diagonal mass matrix.
+        d_L2_proj_matrix_diag[system_name] = std::move(M_vec);
+    }
+
+    IBTK_TIMER_STOP(t_build_diagonal_l2_mass_matrix);
+    return d_L2_proj_matrix_diag[system_name].get();
+}
+
+bool
+FEProjector::computeL2Projection(PetscVector<double>& U_vec,
+                                 PetscVector<double>& F_vec,
+                                 const std::string& system_name,
+                                 const bool consistent_mass_matrix,
+                                 const bool close_U,
+                                 const bool close_F,
+                                 const double tol,
+                                 const unsigned int max_its)
+{
+    IBTK_TIMER_START(t_compute_l2_projection);
+
+    int ierr;
+    bool converged = false;
+
+    if (close_F) F_vec.close();
+    const System& system = d_fe_data->getEquationSystems()->get_system(system_name);
+    const DofMap& dof_map = system.get_dof_map();
+    if (consistent_mass_matrix)
+    {
+        std::pair<PetscLinearSolver<double>*, PetscMatrix<double>*> proj_solver_components =
+            buildL2ProjectionSolver(system_name);
+        PetscLinearSolver<double>* solver = proj_solver_components.first;
+        PetscMatrix<double>* M_mat = proj_solver_components.second;
+        PetscBool rtol_set;
+        double runtime_rtol;
+        ierr = PetscOptionsGetReal(nullptr, "", "-ksp_rtol", &runtime_rtol, &rtol_set);
+        IBTK_CHKERRQ(ierr);
+        PetscBool max_it_set;
+        int runtime_max_it;
+        ierr = PetscOptionsGetInt(nullptr, "", "-ksp_max_it", &runtime_max_it, &max_it_set);
+        IBTK_CHKERRQ(ierr);
+        ierr = KSPSetFromOptions(solver->ksp());
+        IBTK_CHKERRQ(ierr);
+        solver->solve(
+            *M_mat, *M_mat, U_vec, F_vec, rtol_set ? runtime_rtol : tol, max_it_set ? runtime_max_it : max_its);
+        KSPConvergedReason reason;
+        ierr = KSPGetConvergedReason(solver->ksp(), &reason);
+        IBTK_CHKERRQ(ierr);
+        converged = reason > 0;
+    }
+    else
+    {
+        PetscVector<double>* M_diag_vec = buildDiagonalL2MassMatrix(system_name);
+        ierr = VecPointwiseDivide(U_vec.vec(), F_vec.vec(), M_diag_vec->vec());
+        IBTK_CHKERRQ(ierr);
+        converged = true;
+    }
+    if (close_U) U_vec.close();
+    dof_map.enforce_constraints_exactly(system, &U_vec);
+
+    IBTK_TIMER_STOP(t_compute_l2_projection);
+    return converged;
+}
+
+void
+FEProjector::setLoggingEnabled(const bool enable_logging)
+{
+    d_enable_logging = enable_logging;
+}
+
+bool
+FEProjector::getLoggingEnabled() const
+{
+    return d_enable_logging;
+}
```

Comment:
I think 100% of the actual content in this file is copied and pasted so I'm not going to read it :)